src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/InvokerUtil.cs

```diff
@@ -150,8 +168,74 @@ internal static InvokeDelegate GenerateInvokeDelegate(MethodInfo method, out int
 
                     return result;
                 };
+            }
+
+            private static InvokeDelegate GenerateInvokeDelegateInternalWithExpressions(int paramCount, bool returnsValue, int[] inputPos, int[] outputPos, MethodInfo method)
+            {
+                var targetParam = Expression.Parameter(typeof(object), "target");
+                var inputsParam = Expression.Parameter(typeof(object[]), "inputs");
+                var outputsParam = Expression.Parameter(typeof(object[]), "outputs");
+                var paramsLocal = Expression.Variable(typeof(object[]), "paramsLocal");
+                var result = Expression.Variable(typeof(object), "result");
+
+                var methodParam = Expression.Constant(method);
+
+                var assignParamsLocal = Expression.Assign(paramsLocal, Expression.Condition(
+                    Expression.GreaterThan(Expression.Constant(paramCount), Expression.Constant(0)),
+                    Expression.NewArrayBounds(typeof(object), Expression.Constant(paramCount)),
+                    Expression.Constant(null, typeof(object[]))));
+
+                var assignParamsLocalValues = inputPos.Length == 0
+                    ? Expression.Block(Expression.Empty())
+                    : Expression.Block(
+                        Enumerable.Range(0, inputPos.Length).Select(i =>
+                            Expression.Assign(
+                                Expression.ArrayAccess(paramsLocal, Expression.Constant(inputPos[i])),
+                                Expression.ArrayIndex(inputsParam, Expression.Constant(i)))));
+
+                var invokeMethod = Expression.Block(
+                    Expression.Assign(result, Expression.Condition(
+                        Expression.Equal(Expression.Constant(returnsValue), Expression.Constant(true)),
+                        Expression.Call(methodParam, nameof(MethodInfo.Invoke), null, targetParam, paramsLocal),
```

Comment:
It looks like you are calling "Invoke" on the methodInfo instance. So you are still calling `method.Invoke(target, paramsLocal)`. With Expression.Call, you can instead do `Expression.Call(targetParam, methodInfo, arguments)` where `arguments` is an `Expression[]` containing an array populated with Expression instances populated from inputsParam. Basically the loop you unrolled earlier will end up instantiating an array of Expression instead that you will pass here. There is a possibility you will need to emit an `Expression.Cast`, but I hope that's not the case. If it is, then the type you cast to will come from `methodInfo.GetParameters()`. If you do need to do that, make sure you get the type from the parameter array using the position from `inputPos[i]` and not position `i`.

Reply:
done. Please notice that we have to cast in/out around the method invocation with `Expression.Convert`